rflx/pyrflx/typevalue.py

```diff
@@ -260,3 +460,350 @@ def accepted_type(self) -> type:
+                        s = prev_first.value
+                    else:
+                        s = this_first.value
+                    self.set(current_field_name, value[s : s + field_length])
+                    field_first_in_bitstr = s + field_length
+
+            current_field_name = self._next_field(current_field_name)
+
+    def set(self, fld: str, value: Any) -> None:
+
+        # if node is in accessible fields its length and first are known
+        if fld in self.accessible_fields:
+
+            field = self._fields[fld]
+
+            if isinstance(value, Bitstring):
+                if isinstance(field.typeval, OpaqueValue) and not self._has_length(fld):
+                    field.first = self._get_first(fld)
+                    field.typeval.assign_bitvalue(value)
+                    field.length = Number(field.typeval.length)
+                else:
+                    field.first = self._get_first(fld)
+                    field.length = self._get_length(fld)
+                    field.typeval.assign_bitvalue(value)
+            else:
+
+                if not isinstance(value, field.typeval.accepted_type):
+                    raise TypeError(
+                        f"cannot assign different types: {field.typeval.accepted_type.__name__}"
+                        f" != {type(value).__name__}"
+                    )
+
+                # if field is of type opaque and does not have a specified length
+                if isinstance(field.typeval, OpaqueValue) and not self._has_length(fld):
+                    assert isinstance(value, bytes)
+                    field.first = self._get_first(fld)
+                    field.typeval.assign(value)
+                    field.length = Number(field.typeval.length)
+                else:
+                    field.first = self._get_first(fld)
+                    field.length = self._get_length(fld)
+                    field.typeval.assign(value)
+
+        else:
+            raise KeyError(f"cannot access field {fld}")
+
+        if all(
+            [
+                self.__simplified(o.condition) == FALSE
+                for o in self._model.outgoing(model.Field(fld))
+            ]
+        ):
+            self._fields[fld].typeval.clear()
+            print([o.condition for o in self._model.outgoing(model.Field(fld))])
+            raise ValueError("value does not fulfill field condition")
+
+        if isinstance(field.typeval, OpaqueValue) and field.typeval.length != field.length.value:
+            flength = field.typeval.length
+            field.typeval.clear()
+            raise ValueError(f"invalid data length: {field.length.value} != {flength}")
+
+        if isinstance(field.typeval, ArrayValue) and field.typeval.length != field.length.value:
+            flength = field.typeval.length
+            field.typeval.clear()
+            raise ValueError(f"invalid data length: {field.length.value} != {flength}")
```

Comment:
In which cases could be `field.typeval.length != field.length.value`?

Reply:
If an OpaqueValue Field has a defined length of x but less then x bit were assigned (using the set method).

Reply:
Is it needed to be able to set a opaque field only partially? I think it would simplify the code if this special case would be prevented.

Reply:
I am not really sure what you mean. This check should prevent the user from assigning a value which is too short (in case the opaque field has a specified length) if it does not have a specified length `typeval.length == length.value` because of line 583 `field.length = Number(field.typeval.size)`

Reply:
I guess this is related to https://github.com/Componolit/RecordFlux/pull/198#discussion_r408054235?

Reply:
Yes.